src/Workspaces/Remote/Core/ExportProviderBuilder.cs

```diff
@@ -0,0 +1,198 @@
+
+    protected abstract void LogError(string message);
+    protected abstract void LogTrace(string message);
+
+    protected virtual async Task<ExportProvider> CreateExportProviderAsync(CancellationToken cancellationToken)
+    {
+        // Get the cached MEF composition or create a new one.
+        var exportProviderFactory = await GetCompositionConfigurationAsync(cancellationToken).ConfigureAwait(false);
+
+        // Create an export provider, which represents a unique container of values.
+        // You can create as many of these as you want, but typically an app needs just one.
+        var exportProvider = exportProviderFactory.CreateExportProvider();
+
+        return exportProvider;
+    }
+
+    private async Task<IExportProviderFactory> GetCompositionConfigurationAsync(CancellationToken cancellationToken)
+    {
+        // Determine the path to the MEF composition cache file for the given assembly paths.
+        var compositionCacheFile = GetCompositionCacheFilePath();
+
+        // Try to load a cached composition.
+        try
+        {
+            if (File.Exists(compositionCacheFile))
+            {
+                LogTrace($"Loading cached MEF catalog: {compositionCacheFile}");
+
+                CachedComposition cachedComposition = new();
+                using FileStream cacheStream = new(compositionCacheFile, FileMode.Open, FileAccess.Read, FileShare.Read, bufferSize: 4096, useAsync: true);
+                var exportProviderFactory = await cachedComposition.LoadExportProviderFactoryAsync(cacheStream, Resolver, cancellationToken).ConfigureAwait(false);
+
+                return exportProviderFactory;
+            }
+        }
+        catch (Exception ex)
+        {
+            // Log the error, and move on to recover by recreating the MEF composition.
+            LogError($"Loading cached MEF composition failed: {ex}");
+        }
+
+        LogTrace($"Composing MEF catalog using:{Environment.NewLine}{string.Join($"    {Environment.NewLine}", AssemblyPaths)}.");
+
+        var discovery = PartDiscovery.Combine(
+            Resolver,
+            new AttributedPartDiscovery(Resolver, isNonPublicSupported: true), // "NuGet MEF" attributes (Microsoft.Composition)
+            new AttributedPartDiscoveryV1(Resolver));
+
+        var parts = await discovery.CreatePartsAsync(AssemblyPaths, progress: null, cancellationToken).ConfigureAwait(false);
+        var catalog = ComposableCatalog.Create(Resolver)
+            .AddParts(parts)
+            .WithCompositionService(); // Makes an ICompositionService export available to MEF parts to import
+
+        // Assemble the parts into a valid graph.
+        var config = CompositionConfiguration.Create(catalog);
+
+        // Verify we only have expected errors.
+
+        ThrowOnUnexpectedErrors(config, catalog);
+
+        // Try to cache the composition.
+        _ = WriteCompositionCacheAsync(compositionCacheFile, config, cancellationToken).ReportNonFatalErrorAsync();
```

Comment:
is there any concern about two instances of VS loading, both calling into this, and both stomping on each other.  That's a general concern i have here unilaterally.  I'd actually like to talk in person to understand this.